train/compute/python/tools/trace_link.py

```diff
@@ -0,0 +1,309 @@
+import networkx as nx
+from networkx.algorithms import isomorphism
+import json
+from param_bench.train.compute.python.tools.execution_graph import ExecutionGraph
+import sys
+
+
+# Increase recursion limit
+sys.setrecursionlimit(10**6)
+
+
+def read_dictionary_from_json(file_path):
+    with open(file_path, 'r') as file:
+        data = json.load(file)
+        return data
+
+
+# Add and sort ET nodes from the execution graph  
+def collect_nodes(node):
+    def traverse(node):
+        nonlocal nodes
+        nodes.append(node)
+        for child in node.children:
+            traverse(child)
+
+    nodes = []
+    traverse(node)
+    sorted_nodes = sorted(nodes, key=lambda x: x.id)
+    return sorted_nodes
+
+
+class Kineto_node:
+    def __init__(self, name, start, end, id):
+        self.name = name
+        self.start = start
+        self.end = end
+        self.id = id
+        self.children = []
+
+
+# Function to transform your self-defined tree to a directed graph with max depth
+def transform_to_graph_depth(node, max_depth=100):
+    graph = nx.DiGraph()
+    add_node_to_graph_depth(node, graph, max_depth, 1)
+    return graph
+
+
+# Helper function to recursively add nodes and edges to the graph with max depth
+def add_node_to_graph_depth(node, graph, max_depth, cur_depth):
+    graph.add_node(node.id, label=node.name)
+    if cur_depth == max_depth:
+        return
+    for child in node.children:
+        graph.add_node(node.id, label=node.name)
+        add_node_to_graph_depth(child, graph, max_depth, cur_depth + 1)
+
+
+# Custom node comparison function for edit distance
+def node_compare(n1, n2):
+    return n1 == n2
+
+
+# Find the segment that has a length closest to the target
+def find_closest_segment(segs, target_length):
```

Comment:
Curious if the segments are based on iterations in the kineto trace, why are we looking to find the closest length?
Why not match up iterations in ET and kineto trace

Reply:
Currently this assumes ET only contains one iteration and kineto may contain many, so we simply pick the iteration (in kineto) that is most likely to be similar based on number of ops. It is possible to directly match up iterations but this may take much longer time and we can add this later.